src/org/mozilla/javascript/optimizer/BodyCodegen.java

```diff
@@ -3421,26 +3431,55 @@ private static boolean isArithmeticNode(Node node)
             || (type == Token.MUL);
     }
 
-    private void visitArithmetic(Node node, int opCode, Node child,
+    private void visitArithmetic(Node node, int type, Node child,
         Node parent)
     {
         int childNumberFlag = node.getIntProp(Node.ISNUMBER_PROP, -1);
         if (childNumberFlag != -1) {
             generateExpression(child, node);
             generateExpression(child.getNext(), node);
-            cfw.add(opCode);
+
+            switch (type) {
+            case Token.SUB:
+                cfw.add(ByteCode.DSUB);
+                break;
+            case Token.MUL:
+                cfw.add(ByteCode.DMUL);
+                break;
+            case Token.DIV:
+                cfw.add(ByteCode.DDIV);
+                break;
+            case Token.MOD:
+                cfw.add(ByteCode.DREM);
+                break;
+            default:
+                throw Kit.codeBug(Token.typeToName(type));
+            }
         }
         else {
             boolean childOfArithmetic = isArithmeticNode(parent);
             generateExpression(child, node);
             if (!isArithmeticNode(child))
-                addObjectToDouble();
+                addObjectToNumeric();
```

Comment:
```js
(print(1), { valueOf(){ print(2)}}) - (print(3), { valueOf(){ print(4)}})
// actual
// 1
// 2
// 3
// 4

// actual
// 1
// 3
// 2
// 4
```
This is correct.
https://github.com/mozilla/rhino/blob/d02cff98b66bbc9347874720a6d1dbca8195ab8f/src/org/mozilla/javascript/optimizer/BodyCodegen.java#L3460-L3467

Reply:
We need to figure out what to do with `isArithmeticNode`, so we'll leave that for next time.

Reply:
If there are performance regressions here, then I think this is where they'd come, because there is at least a theoretical optimization here that we're partially undoing. If you have any ideas how to keep the old code path in some more cases then I'd be more comfortable. For instance, if "childOfArithmetic" is true at all, can we assume that there is no BigInteger object involved? but I might be OK just doing it this way and then trying to restore the old optimizations later.